packages/vincent-tools/vincent-tool-sdk/src/lib/types.ts

```diff
@@ -0,0 +1,245 @@
+        : PolicyResultAllowNoResult)
+    | (CommitDenyResult extends z.ZodType
+        ? PolicyResultDeny<z.infer<CommitDenyResult>>
+        : PolicyResultDenyNoResult)
+  >;
+}
+
+export interface PolicyWithPrecheckAndCommit<
+  ToolParams extends z.ZodType,
+  UserParams extends z.ZodType,
+  PrecheckAllowResult extends z.ZodType | undefined = undefined,
+  PrecheckDenyResult extends z.ZodType | undefined = undefined,
+  CommitParams extends z.ZodType | undefined = undefined,
+  CommitAllowResult extends z.ZodType | undefined = undefined,
+  CommitDenyResult extends z.ZodType | undefined = undefined,
+  EvalAllowResult extends z.ZodType | undefined = undefined,
+  EvalDenyResult extends z.ZodType | undefined = undefined,
+> extends BasicPolicyDef<
+    ToolParams,
+    UserParams,
+    EvalAllowResult,
+    EvalDenyResult
+  > {
+  precheckAllowResultSchema?: PrecheckAllowResult;
+  precheckDenyResultSchema?: PrecheckDenyResult;
+  precheck: (args: {
+    toolParams: z.infer<ToolParams>;
+    userParams: z.infer<UserParams>;
+  }) => Promise<
+    | (PrecheckAllowResult extends z.ZodType
+        ? PolicyResultAllow<z.infer<PrecheckAllowResult>>
+        : PolicyResultAllowNoResult)
+    | (PrecheckDenyResult extends z.ZodType
+        ? PolicyResultDeny<z.infer<PrecheckDenyResult>>
+        : PolicyResultDenyNoResult)
+  >;
+
+  commitParamsSchema: NonNullable<CommitParams>;
+  commitAllowResultSchema?: CommitAllowResult;
+  commitDenyResultSchema?: CommitDenyResult;
+  commit: (
+    args: CommitParams extends z.ZodType ? z.infer<CommitParams> : never,
+  ) => Promise<
+    | (CommitAllowResult extends z.ZodType
+        ? PolicyResultAllow<z.infer<CommitAllowResult>>
+        : PolicyResultAllowNoResult)
+    | (CommitDenyResult extends z.ZodType
+        ? PolicyResultDeny<z.infer<CommitDenyResult>>
+        : PolicyResultDenyNoResult)
+  >;
+}
+
+// Union type for all policy definitions
+// Do not be alarmed by the `any` types here -- the actual policies are still constrained by the types that make up the union
+export type VincentPolicyDef =
+  | BasicPolicyDef<any, any, any, any>
+  | PolicyWithPrecheck<any, any, any, any, any, any>
+  | PolicyWithCommit<any, any, any, any, any, any, any>
+  | PolicyWithPrecheckAndCommit<any, any, any, any, any, any, any, any, any>;
+
+// Helper type to determine if a policy has commit
+export type HasCommit<P> = P extends { commit: infer CommitFn }
+  ? CommitFn extends Function
+    ? true
+    : false
+  : false;
+
+// Tool supported policy with proper typing
+export type VincentToolSupportedPolicy<
+  ToolParamsSchema extends z.ZodType,
+  PolicyDefType extends VincentPolicyDef,
+> = {
+  policyDef: PolicyDefType;
+  toolParameterMappings: Partial<{
+    [K in keyof z.infer<ToolParamsSchema>]: keyof z.infer<
+      PolicyDefType['toolParamsSchema']
+    >;
+  }>;
+};
+
+// Policy evaluation results with proper typing for commit
+export type VincentPolicyEvaluationResults<
```

Comment:
I think this is also missing the Policy commit execution results

Reply:
We need to add policy commit execution results into the _tool_ response, since they aren't actually part of the evaluation lifecycle. 🤔  I'll look into that tomorrow